x-pack/plugin/index-lifecycle/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleService.java

```diff
@@ -121,7 +123,7 @@ public void applyClusterState(ClusterChangedEvent event) {
         if (event.localNodeMaster()) { // only act if we are master, otherwise
                                        // keep idle until elected
             IndexLifecycleMetadata lifecycleMetadata = event.state().metaData().custom(IndexLifecycleMetadata.TYPE);
-            if (lifecycleMetadata != null && event.changedCustomMetaDataSet().contains(IndexLifecycleMetadata.TYPE)) {
```

Comment:
Hey @jasontedor,

Previously, IndexLifecycleService was not updating its internal state when 
a node was newly-elected master, and there were no changes to policies in the cluster-state.

To fix this, there are a few options, but two that I see are as follows:

1. always attempt to call `policyRegistry.update`, Since the real diff that matters is between the internal state of the registry and the current cluster state.
2. upon un-election, clear the registry, and re-bootstrap it on the first cluster-state listener call that is triggered once it is master.

(2) has the benefit that it makes it clear that this instance is no longer master and therefore should forget about any policies it once had since it is not keeping up to date with updating it anymore for when changes do occur. The downside here is that this can result in weird behavior for when nodes are un-elected, then elected, all before the next state listener callback occurs.

(1) has the benefit that there does not seem to be any edge cases to worry about. The only downside is that the object will be stale until it becomes master again. This should be OK because the cluster-state-applier that will re-update the registry will be called before the cluster-state-listener-callback re-launches the scheduled job and triggers policies.

Are there aspects of this that these thoughts are missing in the story of re-election and state management?

Reply:
update to this inquiry has been discussed in https://github.com/elastic/elasticsearch/issues/32181 and https://github.com/elastic/elasticsearch/pull/32212 was opened to address it